pkg/credentialprovider/apis/criauth/types.go

```diff
@@ -0,0 +1,174 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package criauth
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CRIAuthPluginConfig is the configuration containing information about
+// each CRI auth plugin. Kubelet reads this configuration from disk and enables
+// each plugin as specified by the CRIAuthPlugin type.
+type CRIAuthPluginConfig struct {
+	metav1.TypeMeta
+
+	// plugins is a list of CRI auth plugins that will be enabled by the kubelet.
+	// Multiple plugins may match against a single image, in which case credentials
+	// from all plugins will be returned to the kubelet.
+	Plugins []CRIAuthPlugin
+}
+
+// CRIAuthPlugin is used by the kubelet container runtime to match the
+// image property string (from the container spec) with exec-based credential provider
+// plugins that provide container registry credentials.
+type CRIAuthPlugin struct {
+	// name is the required name of the credential plugin. It must match the name of the
+	// plugin executable as seen by the kubelet. The executable must be in the kubelet's
+	// auth bin directory (set by the --cri-auth-plugin-bin-dir flag).
+	Name string
+
+	// matchImages is a required list of strings used to match against images in order to
+	// determine which plugins should be invoked. If one of the strings matches the
+	// requested image from the kubelet, the plugin will be invoked and given a chance
+	// to provide credentials. Images are expected to contain the registry domain
+	// and URL path.
+	//
+	// Each entry in matchImages is a pattern which can optionally contain a port and a path.
+	// Globs can be used in the domain, but not in the port or the path. Globs are supported
+	// as subdomains like '*.k8s.io' or 'k8s.*.io', and top-level-domains such as 'k8s.*'.
+	// Matching partial subdomains like 'app*.k8s.io' is also supported. Each glob can only match
+	// a single subdomain segment, so *.io does not match *.k8s.io.
+	//
+	// A match exists between an image and a matchImage when all of the below are true:
+	// - Both contain the same number of domain parts and each part matches.
+	// - The URL path of an imageMatch must be a prefix of the target image URL path.
+	// - If the imageMatch contains a port, then the port must match in the image as well.
+	//
+	// Example values of matchImages:
+	//   - 123456789.dkr.ecr.us-east-1.amazonaws.com
+	//   - *.azurecr.io
+	//   - gcr.io
+	//   - *.*.registry.io
+	//   - registry.io:8080/path
+	MatchImages []string
+
+	// defaultCacheDuration is the default duration the plugin will cache credentials in-memory
+	// if a cache duration is not provided in the plugin response. This field is required.
+	DefaultCacheDuration metav1.Duration
+
+	// extraArgs specifies arguments to be passed to the plugin via environment variables.
+	// +optional
+	ExtraArgs []PluginArg
```

Comment:
> This godoc indicates the args are passed in envvars

Sorry this is outdated, it should have said only args.

> client-go exec credential plugins allow control of env and args like this, which seems more straightforward:

Okay I can follow this to be consistent with client-go. FYI @DangerOnTheRanger based on our previous chat. 

> I think we should do something similar (otherwise the default forever has to remain v1alpha1 for compatibility)

Ack.